### DISCUSSION TITLE
CCSDS Support

### DISCUSSION THREAD

**Discussion by satzguyz** [nasa/fprime#1397] [Q&A]
Hello

Does Fprime natively supports CCSDS/AX.25 protocol packets (Framer and De-Framer) or any guide to port it ?

**Comment by JackNWhite** [comment]
If by CCSDS you mean Space Packet protocol, then no. I don't know anything about AX.25. I've been thinking about SPP for F' and it's a bit of a problem, because F' thinks about telemetry especially in quite a different way, with its channels and events.

At each stage of telemetry processing between the source F' component and the downlink framer, some data is serialised and/or modified. In the case of a telemetry channel, the channel ID is added to the component base ID and sent along with the time and the serialised (buffered) payload value when you invoke the channel output port of the component. This is all shunted over to TlmChannel, where is it queued. Before leaving TlmChan, the packet type, full channel ID and send-time are also serialised and buffered. This buffer is then sent to Framer, which uses a framing protocol to add further header information.

The trouble there is that, for putting together a Space Packet primary header, you need an APID (surely related in some way to the channel ID) and packet sequencing information. All that stuff is located in the source component, not the framing protocol.

So, if you want to add SPP in a hacky way, you can deserialise what is already serialised in your framing protocol and recombine it, or you can do it properly and change all the ports related to telemetry and telecommanding to accept all the information required by the framing protocol and TC dispatch. The latter is kind of a job, because from the component-writer's perspective, all that stuff is taken care of behind the scenes in the autocoders. So you have to change them, as well as the ports.

You also need to remove all the non-queuing-related stuff from TlmChan, TlmPacketizer and ActiveLogger, and find a way to make events and channels use the same ID pool, because SPP only has two packet types: TC and TM.

All that reckons without your link protocol, which might require some of that information too, and the possibility of a secondary header within the Space Packet user data.

I have found the devs here to be very helpful, but I don't think there is really much written info about it. You have to set up a component with a channel, an event and a command and then follow all the calls around the code, not forgetting that a lot of it is autocoded.

To do it properly, you will probably need to change every component and port in this diagram
![Fpr_TMTC](https://user-images.githubusercontent.com/10062543/163182765-55f1fffa-6ffe-447f-b64a-7d5893cbdf03.svg)

**Reply by satzguyz** [reply]
Thanks for the update. Any implementation reference for Datalink layer protocols as well ?
![image](https://user-images.githubusercontent.com/103112583/163335166-8b112896-d0c3-4cc8-909b-a9badb183e0a.png)

**Comment by codeflight1** [comment]
I have implemented the CCSDS serialize/deserialize [here](https://github.com/LithiumAerospace/fprime-SpaceWire/tree/master/SpaceWire/CCSDSPacket.cpp), but it currently only takes a Fw::Buffer as data, you would have to connect that to a Framer/Deframer

**Reply by satzguyz** [reply]
Hello @codeflight1 
Does it support Datalink layer as well in framer/de-framer

![image](https://user-images.githubusercontent.com/103112583/163335268-61b86dae-5ce2-4d4b-a1ae-ff734e0ed097.png)

**Reply by codeflight1** [reply]
No, that is not part of SPP specification. SPP only handles application layer.

**Reply by shalevm** [reply]
hey @satzguyz,
i'm wondering if you've had any progress with this issue, as i also need to implement CCSDS protocols, specifically the data-link ones (TM/TC SPACE DATA LINK etc).
any and all implementation references and insights would be much appreciated.
Thanks.
